workbench/src/workbench.c

```diff
@@ -0,0 +1,804 @@
+		{
+			entry = g_ptr_array_index(wb->projects, index);
+			if (entry != NULL && entry->project == address)
+			{
+				return entry->status;
+			}
+		}
+	}
+	return PROJECT_ENTRY_STATUS_UNKNOWN;
+}
+
+
+/** Add a project to the workbench.
+ *
+ * @param wb       The workbench
+ * @param filename Project file
+ * @return TRUE on success, FALSE otherwise
+ *
+ **/
+gboolean workbench_add_project(WORKBENCH *wb, gchar *filename)
+{
+	if (wb != NULL)
+	{
+		GStatBuf buf;
+		WB_PROJECT *project;
+		WB_PROJECT_ENTRY *entry;
+
+		entry = wb_project_entry_new();
+		if (entry == NULL)
+		{
+			return FALSE;
+		}
+		project = wb_project_new(filename);
+		if (project == NULL)
+		{
+			wb_project_entry_free(entry);
+			return FALSE;
+		}
+
+		/* Set entry data:
+		   - absolute and relative filename (relative to workbench file)
+		   - per default use relative path
+		   - check status of project file
+		   - pointer to the project data */
+		entry->abs_filename = g_strdup(filename);
+		entry->rel_filename = get_any_relative_path
+		                          (wb->filename, filename);
+		entry->use_abs = FALSE;
+		entry->project = project;
+		if (g_stat (filename, &buf) == 0)
+		{
+			entry->status = PROJECT_ENTRY_STATUS_OK;
+		}
+		else
+		{
+			entry->status = PROJECT_ENTRY_STATUS_NOT_FOUND;
+		}
+		g_ptr_array_add (wb->projects, entry);
+
+		wb->modified = TRUE;
+		return TRUE;
+	}
+	return FALSE;
+}
+
+
+/** Remove a project from the workbench.
+ *
+ * @param wb      The workbench
+ * @param project The Project
+ * @return TRUE on success, FALSE otherwise
+ *
+ **/
+gboolean workbench_remove_project_with_address(WORKBENCH *wb, WB_PROJECT *project)
+{
+	if (wb != NULL && wb->projects != NULL)
+	{
+		guint index;
+		WB_PROJECT_ENTRY *current;
+
+		for (index = 0 ; index < wb->projects->len ; index++)
+		{
+			current = g_ptr_array_index(wb->projects, index);
+			if (current != NULL && current->project == project)
+			{
+				g_ptr_array_remove_index (wb->projects, index);
+				wb_project_entry_free(current);
+				wb->modified = TRUE;
+				return TRUE;
+			}
+		}
+	}
+	return FALSE;
+}
+
+
+/** Is the file included in the workbench?
+ *
+ * @param wb       The workbench
+ * @param filename The file
+ * @return Address of project in which the file is included.
+ *         NULL if the file is not included in any workbench project.
+ *
+ **/
+WB_PROJECT *workbench_file_is_included (WORKBENCH *wb, gchar *filename)
+{
+	if (wb != NULL)
+	{
+		guint index;
+		WB_PROJECT_ENTRY *current;
+
+		for (index = 0 ; index < wb->projects->len ; index++)
+		{
+			current = g_ptr_array_index(wb->projects, index);
+			if (current != NULL && wb_project_file_is_included(current->project, filename) == TRUE)
+			{
+				return current->project;
+			}
+		}
+	}
+	return NULL;
+}
+
+
+/* Add a workbench bookmark */
+static gboolean workbench_add_bookmark_int(WORKBENCH *wb, gchar *filename)
+{
+	if (wb != NULL)
+	{
+		gchar *new;
+
+		new = g_strdup(filename);
+		if (new != NULL)
```

Comment:
one more place where, if `filename` is not `NULL`, the test is not relevant as `g_strdup()` aborts on allocation failure